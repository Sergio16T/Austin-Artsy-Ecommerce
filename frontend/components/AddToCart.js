import React, { Component } from 'react';
import { Mutation } from 'react-apollo'; 
import gql from 'graphql-tag'; 
import { CURRENT_USER_QUERY } from './User'; 

const ADD_TO_CART_MUTATION = gql`
    mutation ADD_TO_CART_MUTATION($id: ID! ) {
        addToCart(id: $id) {
            id
            quantity
        }
    }
`; 

class AddToCart extends Component {
    render() {
        return (
            <Mutation 
            mutation={ADD_TO_CART_MUTATION} 
            variables={{id: this.props.id}}
            refetchQueries={[{query: CURRENT_USER_QUERY }]}
            >
                {(addToCart, {error, loading}) => 
                    <button type="button" onClick={async () => {
                        const res = await addToCart().catch(err => {
                            alert(err.message); 
                        }); 
                        console.log(res);
                    }}>Add to Cart</button>
                }
            </Mutation>
        );
    }
}

export default AddToCart;