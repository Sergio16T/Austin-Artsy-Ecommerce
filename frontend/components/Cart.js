import React, { Component } from 'react';
import styled from 'styled-components';

const CartWrapper = styled.div`
    position: fixed; 
    height: 100%; 
    width: 400px; 
    transform: ${props => props.isOpen ? "translateX(0)" : "translateX(100%)" }; 
    background: ${props => props.theme.offWhite}; 
    right: 0; 
    top: 0; 
    z-index: 5; 
    border-left: 2px solid ${props => props.theme.purple}; 
    min-width: 500px; 
    transition: all 0.3s;
    overflow: hidden; 
`; 

class Cart extends Component {
    render() {
        return (
            <CartWrapper isOpen={this.props.isOpen}>
                <h2>Cart</h2>
            </CartWrapper>
        );
    }
}

export default Cart;