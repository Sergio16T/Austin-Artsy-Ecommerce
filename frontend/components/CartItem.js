import React from 'react';
import { useMutation } from 'react-apollo'; 
import gql from 'graphql-tag'; 
import styled from 'styled-components';
import RemoveFromCart from './RemoveFromCart'; 
import formatMoney from "../lib/formatMoney";
import { CURRENT_USER_QUERY } from './User';

const UPDATE_CART_ITEM_MUTATION = gql`
    mutation UPDATE_CART_ITEM_MUTATION($updateType: String!, $id: ID!) {
        updateCartItem(updateType: $updateType, id: $id) {
            id
            quantity
        }
    }
`; 

const StyledCartItem = styled.div`
display: grid; 
grid-template-columns: 1fr 2fr .5fr; 
padding: 2rem 0; 
.cartItemImage {
    width: 120px; 
}
@media (max-width: 600px) {
    padding: 1rem 0; 
    .cartItemImage {
        width: 80px; 
    }
}
`; 

const CartItemDescription = styled.div`
    padding-left: 2rem;
.cartitem_title{
    margin-top: 0; 
    font-size: 1.2rem;
}
.priceandquantity {
    display: flex; 
    flex-direction: row; 
}
`; 
const QuantityDiv = styled.div`
    display: flex;
    flex-direction: row; 
    font-size: 1.6rem; 
    padding-left: 2rem; 
    align-items: center; 
.cartItem_Quantity, .minus, .plus {
    padding:  0 .75rem; 
    margin: 0; 
}
.minus, .plus {
    outline: none; 
    border: none; 
    background: transparent; 
    font-size: 1.6rem; 
    cursor: pointer; 
}
.cartitem_price {
    font-size: 1.2rem; 
}
`; 
const CartItem = (props) => {
    const [updateCartItem, { data }] = useMutation(UPDATE_CART_ITEM_MUTATION); 
	const { item } = props.cartItem; 
	return (
		<StyledCartItem>
			<img className="cartItemImage" src={item.image}/> 
			<CartItemDescription>
				<p className="cartitem_title">{item.title}</p>
				<div className="priceandquantity">
					<p className="cartitem_price">{formatMoney(item.price)}</p>
					<QuantityDiv>
						<button className="minus" type="button" onClick={() => {
                            updateCartItem({variables: { updateType: "subtract", id: props.cartItem.id}}); 
                        }} disabled={props.cartItem.quantity === 1}>-</button>
						<p className="cartItem_Quantity">{props.cartItem.quantity}</p>
						<button className="plus" onClick={() => {
                            updateCartItem({variables: { updateType: "add", id: props.cartItem.id}}); 
                        }}>+</button>
					</QuantityDiv>
				</div>
			</CartItemDescription>
			
			<RemoveFromCart id={props.cartItem.id}/>
		</StyledCartItem>
	)
}

export default CartItem; 