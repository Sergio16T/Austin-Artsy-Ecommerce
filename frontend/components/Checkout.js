import React, { Component } from 'react';
import StripeCheckout from 'react-stripe-checkout'; 
import { Mutation } from 'react-apollo'; 
import Router from 'next/router'; 
import Nprogress from 'nprogress'; 
import PropTypes from 'prop-types'; 
import gql from 'graphql-tag'; 
import calculateTotal from '../lib/calculateTotal'; 
import { CURRENT_USER_QUERY } from './User'; 

const CREATE_ORDER_MUTATION = gql`
    mutation CREATE_ORDER_MUTATION($token: String!) {
        createOrder(token: $token) {
            id
						total 
						charge
						items {
							id
							title
						}
        }
    }
`; 

function totalItems(cart) {
    return cart.reduce((startValue, element) => {
        return startValue + element.quantity; 
    }, 0); 
}

class Checkout extends Component {
    onToken = async (res, createOrder) => {
				console.log(res.id); 
				this.props.toggleCart(false);
				const order = await createOrder({
					variables: { token: res.id }
				}).catch(err => {
					console.log(err.message);  
				}); 
				console.log(order); 
				Router.push({
					pathname: "/order", 
					query: {id: order.data.createOrder.id}
				});
			
    }
    render() {
        return (
						<Mutation 
						mutation={CREATE_ORDER_MUTATION}
						refetchQueries={[{ query: CURRENT_USER_QUERY }]}
						>
                {(createOrder, {error, loading}) => 
							<StripeCheckout
							amount={this.props.totalAmount}
							name="Austin's List"
							description={`Order of ${totalItems(this.props.cart)} items`}
							image={this.props.cart.length && this.props.cart[0].item ? this.props.cart[0].item.image : null}
							stripeKey="pk_test_lUHGmLxt0KGqdsIN20tuXiVw00Q2OwgjiR"
							currency="USD"
							email={this.props.email}
							token={(res) => this.onToken(res, createOrder)}
							>{this.props.children}</StripeCheckout>
            }
            </Mutation>
        );
    }
}

export default Checkout;