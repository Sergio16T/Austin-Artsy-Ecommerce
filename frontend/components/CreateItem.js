import React, { Component } from 'react';
import { Query, Mutation } from 'react-apollo'; 
import Router from 'next/router'; 
import gql from'graphql-tag'; 
import { StyledFormWrapper, StyledForm} from './styles/FormStyles'; 
import Spinner from './Spinner'; 
import { ALL_ITEMS_QUERY } from './Items'; 
import { PAGINATION_QUERY } from './Pagination'; 

const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION(
        $title: String! 
        $description: String!
        $price: Int! 
        $image: String!
        $largeImage: String!
        ) {
        createItem(
            title: $title, 
            description: $description, 
            price: $price
            image: $image
            largeImage: $largeImage 
            ) {
            id
        }
    }
`; 
class CreateItem extends Component {
    state = {
        title: '', 
        price: 0, 
        description: '',
        image: '',
        largeImage: '',
        errorMessage: '',
        uploadError: ''
    }
    uploadFile = async (e) => {
        console.log('uploading file'); 
        this.setState({ spinner: true }); 
        const files = e.target.files; 
        const data = new FormData(); 
        data.append('file', files[0]); 
        data.append('upload_preset', 'AustinArts'); 

        const res = await fetch('https://api.cloudinary.com/v1_1/dddnhychw/image/upload', {
            method: 'POST', 
            body: data  
        }); 
        const file = await res.json(); 
        console.log(file); 

        try {
            this.setState({
                image: file.secure_url, 
                largeImage: file.eager[0].secure_url,
                spinner: false
            }); 
        }
        catch(err) {
            this.setState({uploadError: err.message}); 
        }
    }
    handleInput = (e) => {
        let regEx = /^\d+(\.\d{0,2})?$/; 
        const { value, type, name } = e.target; 
        let val = type === 'number' && value.length ? parseFloat(value) : value;  
        // console.log(regEx.test(val)); 
        try {
            if ((typeof val === 'number') && !regEx.test(val)) {
                throw new Error('Only two decimal points please'); 
            }
        }
        catch(err) {
            console.log(err.message);
            this.setState({
                errorMessage: err.message
            }); 
        }
        if ((typeof val === 'number') && regEx.test(val)) {
            this.setState({
                errorMessage: ''
            }); 
        }
        this.setState({
            [name]: val
        }); 
    }
    render() {
        let regEx2= /^\d+(\.\d{0,1})$/; 
        const { errorMessage } = this.state; 
        const setPrice = regEx2.test(this.state.price) ? this.state.price.toFixed(2) : this.state.price; 
        return (
            <StyledFormWrapper>
                <div className="formContainer">
                    <Query query={PAGINATION_QUERY}>
                        {({data, error, loading}) => {
                            if(loading) return null; 
                            const count = data.itemsConnection.aggregate.count; 
                            // console.log(count); 
                            let lastOnPage  = count/4 === 0; 
                            let page = Math.ceil((count+1)/4) 
                            // console.log(page); 
                            let skip = page * 4 - 4;
                            // console.log(skip); 
                            return (
                        <Mutation 
                        mutation={CREATE_ITEM_MUTATION} 
                        variables={{...this.state, price: setPrice * 100}}
                        refetchQueries={[
                            { query: PAGINATION_QUERY }, 
                            { query: ALL_ITEMS_QUERY, variables: {skip: skip} }, 
                            { query: ALL_ITEMS_QUERY, variables: {skip: (page-1) * 4 - 4 }}
                        ]}
                        >
                            {(createItem, {loading, error}) => (
                            <StyledForm onSubmit={ async (e)=> {
                                e.preventDefault(); 
                                this.setState({ spinner: true });
                                let res = await createItem().catch(err => {
                                    if(err) {
                                        this.setState({ 
                                            spinner: false, 
                                            error: true
                                         }); 
                                    } else {
                                        this.setState({
                                            error: false
                                        }); 
                                    }  
                                }); 
                                this.setState({ spinner: false }); 
                                console.log(res); 
                                if(!this.state.error) {
                                    Router.push({
                                        pathname: "/items", 
                                        query: {page: page}
                                        // query: {id: res.data.createItem.id}
                                    }); 
                                }
                            }} 
                            errorMessage = {errorMessage}
                            >
                                <Spinner spinner={this.state.spinner}/>
                                {error && <p className="errorMessage">{error.message.replace("GraphQL error:", "")}</p>}
                                <fieldset disabled={loading} aria-busy={loading}>
                                    <label htmlFor="file">
                                        Image
                                        <input
                                        type ="file" 
                                        id ="file" 
                                        name="file" 
                                        placeholder="upload an image" 
                                        onChange={this.uploadFile}
                                        />
                                    </label>
                                    {this.state.largeImage && <img width="250" src={this.state.image} alt={this.state.description}></img>}
                                    <label htmlFor="title">
                                        Title
                                        <input
                                        type ="text" 
                                        id ="title" 
                                        name="title" 
                                        placeholder="title" 
                                        value={this.state.title}
                                        onChange={this.handleInput}
                                        required  
                                        />
                                    </label>
                                    <label htmlFor="price" id="priceLabel">
                                        {errorMessage ? errorMessage : 'Price'}
                                        <input
                                        type ="number" 
                                        id ="price" 
                                        name="price" 
                                        placeholder="price" 
                                        value={this.state.price}
                                        step=".01"
                                        onChange={this.handleInput}
                                        required  
                                        />
                                    </label>
                                    <label htmlFor="description">
                                        Description
                                        <textarea
                                        id ="description" 
                                        name="description" 
                                        placeholder="Enter a Description" 
                                        value={this.state.description}
                                        onChange={this.handleInput}
                                        required  
                                        />
                                    </label>
                                    <button type="submit" disabled={errorMessage ? true : false} aria-disabled={errorMessage ? true : false}> Submit </button>
                                </fieldset>
                            </StyledForm>
                        )}
                        </Mutation> 
                    )}}
                </Query>
                </div>
            </StyledFormWrapper>
        );
    }
}

export default CreateItem;