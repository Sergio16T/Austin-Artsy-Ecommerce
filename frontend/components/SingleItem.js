import React, { Component } from 'react';
import { Query } from 'react-apollo'; 
import {SINGLE_ITEM_QUERY} from './UpdateItem'; 
import styled from 'styled-components'; 
import formatMoney from '../lib/formatMoney'; 

const ItemWrapper = styled.div`
    position: relative; 
    padding-top: 100px; 
    /* position: relative; 
    background: #FAFAFA; */
    padding: 15% 35px;  
    background: linear-gradient(to Bottom, rgba(10,10,54,1) 100px, white 0%);  
    @media (max-width: 800px) {
        /* background: linear-gradient(to Bottom, rgba(10,10,54,1) 10%, white 0%);  
        padding: 15% 0;  */
    }
`;
const SingleItemStyles = styled.div`
    max-width: 1200px;
    box-shadow: ${props => props.theme.bs}; 
    display: grid; 
    grid-template-columns: 1fr 1fr; 
    grid-column-gap: 50px; 
    grid-auto-flow: column; 
    min-height: 800px; 
    background: white; 
    .gridContainer {
        display: flex; 
        position: relative; 
        .gridBox {
            height: 420px; 
        }
    }
    .itemBox {
        display: flex;
        text-align: center;
        flex-direction: column;
    }
    .itemTitle {
        font-size: 2rem; 
    }
    .itemDesc {
        font-size: 1.2rem; 
    }
    img {
        width: 100%; 
        height: 100%; 
        object-fit: contain; 
    }
    .details {
        margin: 3rem; 
        font-size: 2rem; 
    }
    @media (max-width: 900px) {
        /* margin-top: 20%;  */
        grid-template-columns: 1fr;
        grid-template-rows: .5fr 1fr; 
        .gridBox {
            height: auto; 
        }
    }
`;

class SingleItem extends Component {
    render() {
        return (
            <Query query={SINGLE_ITEM_QUERY} variables={{id: this.props.id}}>
                {({data, error, loading}) => {
                if(loading) return null; 
                if(error) return <p>{error.message}</p>
                if(!data.item) return <p>Item not found for {this.props.id}</p>
                const item = data.item; 
                if(data) console.log(data); 
                return (
                    <ItemWrapper>
                        <SingleItemStyles>
                            <div className="gridContainer">
                                <div className="gridBox">
                                <img src={data.item.largeImage} alt={data.item.title}/>
                                </div>
                            </div>
                            <div className="itemBox">
                               <h2 className="itemTitle">Viewing {data.item.title}</h2> 
                               <p className="itemDesc">{formatMoney(data.item.price)}</p>
                            </div>
                        </SingleItemStyles>
                    </ItemWrapper>
            
                )}}
            </Query>
        );
    }
}

export default SingleItem;

