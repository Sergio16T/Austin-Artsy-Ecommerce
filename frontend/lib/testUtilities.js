import casual from 'casual'; 
//.seed() is to provide ability to use a specific seed in order to get a repeatable random sequence for our testing purposes
casual.seed(111); 

const mockItem = () => ({
    // __typename: 'Item',
    id: '123',
    price: 5000,
    user: null,
    image: 'dog.jpg',
    title: 'dogs are the best',
    description: 'dogs description',
    largeImage: 'dog-large.jpg',
}); 

const mockUserWithItemsInCart = () =>  ({
    __typename: 'User', 
    id: "123", 
    name: casual.name, 
    email: casual.email, 
    permissions: ["ADMIN"], 
    cart: [{
        id: 'abc', 
        quantity: 2, 
        item: {
            id: 'dgh', 
            price: 15000, 
            image: 'www.dog.image.com',
            title: "puppy"
        }
        },
    ]
}); 
const mockUser = () => ({
    __typename: 'User', 
    id: "123", 
    name: casual.name, 
    email: casual.email, 
    permissions: ['ADMIN'], 
    cart: [],
    orders: []
}); 
const mockOrderItem = () => ({
    __typename: 'OrderItem',
    id: casual.uuid,
    image: `${casual.word}.jpg`,
    title: casual.words(),
    price: 4234,
    quantity: 1,
    description: casual.words(),
  });
  
  const mockOrder = () => ({
    __typename: 'Order',
    id: 'ord123',
    charge: 'ch_123',
    total: 40000,
    items: [mockOrderItem(), mockOrderItem()],
    createdAt: '2018-04 - 06T19: 24: 16.000Z',
    user: mockUser(),
  });
  
  const mockCartItem = overrides => ({
    __typename: 'CartItem',
    id: 'omg123',
    quantity: 3,
    item: mockItem(),
    user: mockUser(),
    ...overrides,
  });
//mock LocalStorage 
class MockLocalStorage {
    constructor(){
        this.store={}; 
    }

    clear() {
        this.store = {}; 
    }
    getItem(key) {
        return this.store[key] || null; 
    }
    setItem(key, value) {
        this.store[key] = value.toString(); 
    }
    removeItem(key) {
        delete this.store[key]; 
    }
}

export { 
    mockItem, 
    MockLocalStorage, 
    mockUser, 
    mockCartItem, 
    mockOrderItem, 
    mockOrder }; 