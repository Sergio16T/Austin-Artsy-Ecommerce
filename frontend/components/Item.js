import React, { Component } from 'react'; 
import Link from 'next/link'; 
import { StyledItem } from './styles/ItemStyles'; 
import styled from 'styled-components'; 
import formatMoney from '../lib/formatMoney';
import DeleteItem from "./DeleteItem"; 
import AddToCart from './AddToCart'; 

const Title = styled.div`
    a { 
        font-size: 1.5rem; 
    }
`; 
const ItemButtons = styled.div`
    display: grid; 
    grid-template-columns: 1fr 1fr 1fr; 
    border-top: 1px solid #E1E1E1;
    border-bottom: 1px solid #E1E1E1; 
    grid-gap: 1px; 
    background: #E1E1E1;
    button { 
        background: white; 
        cursor: pointer; 
        padding: 10px; 
        border: none; 
        outline: none; 
    }
    #editButton {
        background: white; 
        padding: 10px; 
    }
`; 
class Item extends Component {
   constructor(props) {
       super(props); 
   }
    render() {
        const { item } = this.props; 
        return (
            <StyledItem>
                {item.image && <img src={item.image[0]} alt={item.title}/>}
                <Title>
                    <Link href={{
                        pathname: "/item", 
                        query: {id: item.id}
                    }}>
                        <a>{item.title}</a>
                    </Link>
                </Title>
                <p>{item.description}</p>
                <p>{formatMoney(item.price)}</p>
                <ItemButtons>
                    <Link href={{
                        pathname: '/update', 
                        query: {id: item.id}
                    }}>
                        <a id="editButton"> Edit </a>
                    </Link>
                    <AddToCart id={item.id}/>
                    <DeleteItem id={item.id} page={this.props.page}>Delete</DeleteItem>
                </ItemButtons>
            </StyledItem>      
        );
    }
}

export default Item;
export { ItemButtons }