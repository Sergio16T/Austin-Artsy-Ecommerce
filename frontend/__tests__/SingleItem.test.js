import { mount } from 'enzyme'; 
import { shallowToJson } from 'enzyme-to-json'; 
import SingleItem, { SINGLE_ITEM_QUERY } from '../components/SingleItem'; 
import { MockedProvider } from '@apollo/react-testing'; 
import { mockItem } from '../lib/testUtilities'; 
import { ItemButtons } from '../components/Item';
import { act } from 'react-dom/test-utils'; 
import wait from 'waait'; 

// const wait = (amount = 0) => new Promise(resolve => setTimeout(resolve, amount)); 

describe('<SingleItem/>', () => {
    it ('renders with correct data', async ()=> {
        const mocks = [{
            // when someone mackes a request with this combo 
            request: { query: SINGLE_ITEM_QUERY, variables: {id: '123' }},
            // return this fake data 
            result: { data: {
                item: mockItem()
            }, 
          }, 
        }]; 
     
        const wrapper = mount(
                <MockedProvider addTypename={false} mocks={mocks}>
                     <SingleItem id='123'/>
                </MockedProvider>
            ); 
    
			
       await wait(); 
 
        act(() => {
            wrapper.update(); 
            // console.log(wrapper.debug()); 
        }); 
        expect(shallowToJson(wrapper.find('h2'))).toMatchSnapshot(); 
        expect(shallowToJson(wrapper.find('img'))).toMatchSnapshot(); 
        expect(shallowToJson(wrapper.find('p'))).toMatchSnapshot(); 
				
		}); 
		it('Errors with a not found item', async ()=> {
			const mocks = [{
				request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' }}, 
				result: { 
					errors: [{ message : 'Item not found '}]
				}
			}]; 

			const wrapper = mount(
				<MockedProvider addTypename={false} mocks={mocks}>
						 <SingleItem id='123'/>
				</MockedProvider>
		); 
		 await wait(); 
		 wrapper.update(); 
		//  console.log(wrapper.debug());
		 const item = wrapper.find('[data-test="graphql-error"]'); 
		 console.log(item.debug()); 
		 expect(item.text()).toContain("Item not found"); 
		 expect(shallowToJson(item)).toMatchSnapshot(); 
		}); 

})