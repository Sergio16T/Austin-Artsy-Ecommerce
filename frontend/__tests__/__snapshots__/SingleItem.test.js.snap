// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem/> Errors with a not found item 1`] = `
<p
  data-test="graphql-error"
>
   Item not found 
</p>
`;

exports[`<SingleItem/> renders with correct data 1`] = `
<h2
  className="itemTitle"
>
  Viewing 
  dogs are the best
</h2>
`;

exports[`<SingleItem/> renders with correct data 2`] = `
<img
  alt="dogs are the best"
  src="dog.jpg"
/>
`;

exports[`<SingleItem/> renders with correct data 3`] = `
<p
  className="itemDesc"
>
  $50.00
</p>
`;
